app/extensions/email/__init__.py

```diff
@@ -105,63 +157,173 @@ def template(self, template, **kwargs):
+            return
+        self.body = self._try_templates('txt')
+
+    def _render_html(self):
+        if self.html:
+            return
+
+        # Render raw HTML template with Jinja2
+        self.raw_html = self._try_templates('html')
+
+        # Run Premailer
+        attempt = 0
+        while attempt <= 3:
+            attempt += 1
+            try:
+                transformed_html = pmail.transform(self.raw_html)
+                break
+            except (cssutils.prodparser.Missing):
+                pass
+
+        # Strip out unused leftover CSS and minify before sending
+        assert NEWLINE_TEMP_CODE not in transformed_html
+        transformed_html_ = transformed_html.replace('\n', NEWLINE_TEMP_CODE)
+        minified_css_html_ = re.sub(
+            r'<style type="text/css">.*</style>', '', transformed_html_
+        )
+        minified_css_html = minified_css_html_.replace(NEWLINE_TEMP_CODE, '\n')
+        minified_html = htmlmin.minify(
+            minified_css_html,
+            remove_comments=True,
+            remove_empty_space=True,
+            remove_all_empty_space=True,
+        )
+
+        # Add web fonts
+        webfonts = [
+            '<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Assistant:200|Chango|Molle:400i&display=swap">',
+            '<link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">',
+            '<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Code+Pro&display=swap">',
+        ]
+        webfonts_html = ''.join(webfonts)
+        minified_html = minified_html.replace(
+            '</head>', '%s</head>' % (WEBFONTS_PLACEHOLDER_CODE,)
+        )
+        final_html = minified_html.replace(WEBFONTS_PLACEHOLDER_CODE, webfonts_html)
+        with open('email.latest.html', 'w') as temp:
+            temp.write(final_html)
+        self.html = final_html
+
     def attach(self, filepath, atatchment_name, attachment_type='image/png'):
         with current_app.open_resource(filepath) as asset:
             self.attach(atatchment_name, attachment_type, asset.read())
 
         return self
 
+    # note: in order to not get complex and have to break one Email up into multiple, we just use the first language
+    #   we find on a recipient; TODO develop a potential MultiLanguageEmail which is acually a (potential) list of Emails
+    def set_language(self):
+        if self.language or not self.recipients:
+            return
+
+        from app.modules.site_settings.models import SiteSetting
+        from app.modules.users.models import User
+
+        self.resolve_recipients()
+        for recip in self._original_recipients:
+            if isinstance(recip, User):
+                self.language = recip.get_preferred_langauge()
+                if self.language:
+                    return
+        self.language = SiteSetting.get_string('preferred_language', 'en_us')
+
+    def resolve_recipients(self):
```

Comment:
In the __init__ function it does other things setting up recipients which I think is then over written here. Wouldn't it be better to have the recipients set up in the __init__ function once and move this functionality there?